src/main/java/io/debezium/connector/jdbc/JdbcSinkConnectorTask.java

```diff
@@ -0,0 +1,142 @@
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.persistence.EntityManager;
+import javax.xml.bind.JAXBException;
+
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.apache.kafka.connect.sink.SinkTask;
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.Transaction;
+import org.hibernate.boot.jaxb.hbm.spi.JaxbHbmHibernateMapping;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The main task executing streaming from sink connector.
+ * Responsible for lifecycle management of the streaming code.
+ *
+ * @author Hossein Torabi
+ *
+ */
+public class JdbcSinkConnectorTask extends SinkTask {
+    private static enum State {
+        RUNNING,
+        STOPPED;
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(JdbcSinkConnectorTask.class);
+
+    private final AtomicReference<State> state = new AtomicReference<State>(State.STOPPED);
+    private final ReentrantLock stateLock = new ReentrantLock();
+    private final Map<String, Schema> topicSchema = new HashMap<>();
+
+    private HibernateJAXBRecordMapper jaxbRecordMapper;
+    private JdbcSinkConnectorConfig config;
+
+    private SessionFactory sessionFactory;
+    private EntityManager entityManager;
+    private Session session;
+
+    @Override
+    public String version() {
+        return Module.version();
+    }
+
+    @Override
+    public void start(Map<String, String> props) {
+        stateLock.lock();
+
+        try {
+            if (!state.compareAndSet(State.STOPPED, State.RUNNING)) {
+                LOGGER.info("Connector has already been started");
+                return;
+            }
+
+            config = new JdbcSinkConnectorConfig(props);
+            config.validate();
+
+            jaxbRecordMapper = new HibernateJAXBRecordMapper(config);
+            buildSessionFactory();
+
+        }
+        finally {
+            stateLock.unlock();
+        }
+    }
+
+    @Override
+    public void put(Collection<SinkRecord> records) {
+        records.forEach((SinkRecord record) -> {
+            String topic = record.topic();
+            Schema schema = record.valueSchema();
+            if (!topicSchema.containsKey(topic) || !topicSchema.get(topic).equals(schema)) {
+                topicSchema.put(topic, schema);
+                JaxbHbmHibernateMapping mapping = jaxbRecordMapper.getJaxbHbmHibernateMapping(topic, schema);
+                try {
+                    persistJaxbHbmHibernateMapping(topic, mapping);
+                }
+                catch (JAXBException | FileNotFoundException e) {
+                    throw new RuntimeException("Can not persists new schema %s", e);
+                }
+                buildSessionFactory();
+            }
+
+            Transaction transaction = session.beginTransaction();
```

Comment:
It probably makes more sense to have one transaction for the entire incoming batch rather than per-record. Also the session should be recreated per batch (otherwise, Hibernate's first-level cache will keep growing indefinitely), i.e. it should be local variable in this method rather than a field.